cpp/src/io/parquet/page_enc.cu

```diff
@@ -0,0 +1,1544 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "parquet_gpu.h"
+#include <io/utilities/block_utils.cuh>
+
+namespace cudf {
+namespace io {
+namespace parquet {
+namespace gpu {
+
+#define INIT_HASH_BITS              12
+
+struct frag_init_state_s
+{
+    EncColumnDesc col;
+    PageFragment frag;
+    uint32_t total_dupes;
+    volatile uint32_t scratch_red[32];
+    uint32_t dict[MAX_PAGE_FRAGMENT_SIZE];
+    union {
+        uint16_t u16[1 << (INIT_HASH_BITS)];
+        uint32_t u32[1 << (INIT_HASH_BITS - 1)];
+    } map;
+};
+
+#define LOG2_RLE_BFRSZ  9
+#define RLE_BFRSZ       (1 << LOG2_RLE_BFRSZ)
+#define RLE_MAX_LIT_RUN 0xfff8  // Maximum literal run for 2-byte run code
+
+struct page_enc_state_s
+{
+    uint8_t *cur;               //!< current output ptr
+    uint8_t *rle_out;           //!< current RLE write ptr
+    uint32_t rle_run;           //!< current RLE run
+    uint32_t run_val;           //!< current RLE run value
+    uint32_t rle_pos;           //!< RLE encoder positions
+    uint32_t rle_numvals;       //!< RLE input value count
+    uint32_t rle_lit_count;
+    uint32_t rle_rpt_count;
+    volatile uint32_t rpt_map[4];
+    volatile uint32_t scratch_red[32];
+    EncPage page;
+    EncColumnChunk ck;
+    EncColumnDesc col;
+    gpu_inflate_input_s comp_in;
+    gpu_inflate_status_s comp_out;
+    uint16_t vals[RLE_BFRSZ];
+};
+
+
+/**
+ * @brief Return a 12-bit hash from a byte sequence
+ */
+inline __device__ uint32_t nvstr_init_hash(const uint8_t *ptr, uint32_t len)
+{
+    if (len != 0)
+    {
+        return (ptr[0] + (ptr[len - 1] << 5) + (len << 10)) & ((1 << INIT_HASH_BITS) - 1);
+    }
+    else
+    {
+        return 0;
+    }
+}
+
+inline __device__ uint32_t uint32_init_hash(uint32_t v)
+{
+    return (v + (v >> 16)) & ((1 << INIT_HASH_BITS) - 1);
```

Comment:
Surprised to see such a simple hash function.  Are there concerns the hash distribution will lead to a large number of collisions?

Reply:
Not a big concern at this stage, as it's only trying to do early duplicate elimination. It seemed to work relatively well for ints and floats (unless we expect to have lots of values where the low 12-bits are identical). The final elimination (where collisions matter more) in page_dict uses 16-bit, but the same hash method (does something more fancy for strings, though). The dictionary size being limited to 64K entries somewhat limits the worst-case (perf is typically dominated by strings)

Reply:
In this early stage there are only 5000 values being hashed with a 12-bit hash. If they all end up with the same hash, it could result in failure to detect most duplicates, which could result in dictionary encoding not being used when it should -> would be good to keep an eye open for  datasets where we end up with significantly worse compression than the CPU parquet-cpp writer (or unusually poor dictionary perf if we also get significant hash collisions in the second stage).

Reply:
Thinking about it, it's trivial to just at least not drop bits 12..15 and 28..31, which would at least avoid floats x and -x having the same hash...